OptimoveSDK/optimove-sdk/src/main/java/com/optimove/android/embeddedmessaging/OptimoveEmbeddedMessaging.java

```diff
@@ -0,0 +1,433 @@
+        if (shared == null) {
+            throw new IllegalStateException("OptimoveEmbeddedMessagesModule is not initialized");
+        }
+        return shared;
+    }
+
+    public static void initialize() {
+        shared = new OptimoveEmbeddedMessaging();
+        executorService = Executors.newSingleThreadExecutor();
+    }
+
+    public enum ResultType {
+        SUCCESS,
+        ERROR_USER_NOT_SET,
+        ERROR_CONFIG_NOT_SET,
+        ERROR
+    }
+
+    public interface EmbeddedMessagesGetHandler {
+        void run(ResultType result, @Nullable List<Container> containers);
+    }
+
+    public interface EmbeddedMessagesSetHandler {
+        void run(ResultType result);
+    }
+
+    public class EmbeddedMessagingResult {
+        private ResultType result;
+        private List<Container> containers;
+
+        public EmbeddedMessagingResult(ResultType result, List<Container> containers) {
+            this.result = result;
+            this.containers = containers;
+        }
+
+        public ResultType getResult() {
+            return this.result;
+        }
+
+        public List<Container> getContainers() {
+            return this.containers;
+        }
+    }
+
+    public void getEmbeddedMessagesAsync(ContainerMessageRequest[] requestBody, @NonNull EmbeddedMessagesGetHandler embeddedMessagesGetHandler) {
+        EmbeddedMessagingConfig config = Optimove.getConfig().getEmbeddedMessagingConfig();
+        if (config == null) {
+            Log.e(TAG, "Embedded messaging config is not set");
+            handler.post(() -> embeddedMessagesGetHandler.run(ResultType.ERROR_CONFIG_NOT_SET, null));
+            return;
+        }
+
+        UserInfo userInfo = Optimove.getInstance().getUserInfo();
+        String userId = userInfo.getUserId() == null ? userInfo.getVisitorId() : userInfo.getUserId();
+
+        if (userId == null) {
+            Log.w(TAG, "Customer/Visitor ID is not set");
+            handler.post(() -> embeddedMessagesGetHandler.run(ResultType.ERROR_USER_NOT_SET, null));
+            return;
+        }
+        Runnable task = new GetEmbeddedMessagesRunnable(config, userId, embeddedMessagesGetHandler, requestBody);
+        executorService.submit(task);
+    }
+
+    public void getEmbeddedMessagesAsync(@NonNull EmbeddedMessagesGetHandler embeddedMessagesGetHandler) {
+        getEmbeddedMessagesAsync(new ContainerMessageRequest[]{}, embeddedMessagesGetHandler);
+    }
+
+    public void deleteEmbeddedMessageAsync(String Id, @NonNull EmbeddedMessagesSetHandler embeddedMessagesDeleteHandler) {
+        EmbeddedMessagingConfig config = handleConfigForAsyncSetCall(embeddedMessagesDeleteHandler);
+        if (config == null) return;
+
+        Runnable task = new DeleteEmbeddedMessagesRunnable(config, Id, embeddedMessagesDeleteHandler);
+        executorService.submit(task);
+    }
+
+    public void reportClickMetricAsync(
+            EmbeddedMessageMetricsRequest metrics,
```

Comment:
As discussed, let's take messages in here instead to hide the impl details & future-proof.